src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -2,76 +2,70 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
-using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using ICSharpCode.Decompiler.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CodeCleanup;
 using Microsoft.CodeAnalysis.CodeGeneration;
 using Microsoft.CodeAnalysis.CSharp.CodeGeneration;
 using Microsoft.CodeAnalysis.CSharp.Extensions;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Editor.BackgroundWorkIndicator;
 using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
-using Microsoft.CodeAnalysis.Editor.Shared.Options;
+using Microsoft.CodeAnalysis.ErrorReporting;
 using Microsoft.CodeAnalysis.EventHookup;
 using Microsoft.CodeAnalysis.Formatting;
-using Microsoft.CodeAnalysis.Internal.Log;
 using Microsoft.CodeAnalysis.LanguageService;
-using Microsoft.CodeAnalysis.Options;
-using Microsoft.CodeAnalysis.Rename;
 using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Microsoft.CodeAnalysis.Simplification;
 using Microsoft.CodeAnalysis.Text;
 using Microsoft.CodeAnalysis.Text.Shared.Extensions;
 using Microsoft.VisualStudio.Commanding;
 using Microsoft.VisualStudio.Text;
-using Microsoft.VisualStudio.Text.Editor;
 using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Threading;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.EventHookup;
 
 internal partial class EventHookupCommandHandler : IChainedCommandHandler<TabKeyCommandArgs>
 {
-    public void ExecuteCommand(TabKeyCommandArgs args, Action nextHandler, CommandExecutionContext context)
+    public CommandState GetCommandState(TabKeyCommandArgs args, Func<CommandState> nextHandler)
     {
         _threadingContext.ThrowIfNotOnUIThread();
-        if (!_globalOptions.GetOption(EventHookupOptionsStorage.EventHookup))
+        if (EventHookupSessionManager.CurrentSession != null)
         {
-            nextHandler();
-            return;
+            return CommandState.Available;
         }
-
-        if (EventHookupSessionManager.CurrentSession == null)
+        else
         {
-            nextHandler();
-            return;
+            return nextHandler();
         }
-
-        // Handling tab is currently uncancellable.
-        HandleTabWorker(args.TextView, args.SubjectBuffer, nextHandler, CancellationToken.None);
     }
 
-    public CommandState GetCommandState(TabKeyCommandArgs args, Func<CommandState> nextHandler)
+    public void ExecuteCommand(TabKeyCommandArgs args, Action nextHandler, CommandExecutionContext context)
     {
         _threadingContext.ThrowIfNotOnUIThread();
-        if (EventHookupSessionManager.CurrentSession != null)
+        if (!TryExecuteCommand(args, nextHandler))
```

Comment:
Execute first calls into the up front synchronous portion of TryExecute.  if that part fails, we ensure we're reset back to normal and pass the tab into the editor.